README.md
<img src="Docs/Resources/LandOS-JPEG-Dark.jpg" alt="LandOS Logo" height="500px"><br />
16 bit operating system written completely from scratch in Assembly x86
## How do I use this
<img src="Docs/Resources/winlogo.png" alt="Windows Logo" width="12px"> Instructions for windows: <br />
<b>REQUIREMENTS</b><br />
<ul>
  <li>NASM for Windows</li>
  <li>QEMU</li>
</ul>
The source code is very easy to compile.  First you need to go to the directory that nasm works in (usually C:\Program Files\NASM) and then you want to execute the following code: <br /><br />

```> nasm boot_sect.asm -f bin -o boot_sect.bin``` <br />
<i>unless "boot_sect.asm" is in the NASM directory, you'll want to put the complete address.  Same goes for where you want your output file to be.</i>

Next you'll want to run it with QEMU.  You can try other programs but QEMU is the only one that has worked so far.  Make sure to move your compiled bootloader (.bin) into your qemu folder (usually C:\Program Files\qemu)<br />
Then, assuming you have qemu all set up, you can use the following command: <br />

```> qemu-system-x86_64.exe boot_sect.bin``` <br />
If your compiled bootloader isn't named "boot_sect" you would replace it with whatever the file is named.<br /><br />
<img src="Docs/Resources/551CEAE2-F788-4D9B-8268-485204182B89.png" alt="Termux Logo" width="14px"> Instructions for Termux:<br />
<b>REQUIREMENTS</b><br />
<ul>
  <li>TigerVNC</li>
  <li>VNC VIEWER</li>
  <li>QEMU(x86_64)</li>
  <li>OPTIONAL: NASM</li>
</ul>
For those who want to use emulate LandOS on a Termux-friendly device will need a few different resources.  This requires a good couple of steps but is a lot easier the second time.  First you're going to want to install the neccessary packages.  We'll install Termux's X11 repo first<br /><br />

```> pkg install x11-repo```<br />
Once thats finished, we'll install TigerVNC<br /><br />
```> pkg install TigerVNC```<br />

Finally, we want to install QEMU (Specifically x86_64)<br />< br/>

```> pkg install qemu-system-x86_64```<br />

Now we Have our neccessary packages and we can begin setting up our enviroment.  First you'll need to be sure to navigate to the directory where the image you want to use is stored.  Then you'll want to start rhe server using:<br />

```> vncserver -localhost```<br/>
If this is TigerVNC's first time setting up a VNC server, you'll be prompted to set up a VNC server password.  Once it is finished, assuming you only have one enviroment running, you can export the display to the VNC server using:<br /><br />

```> export DISPLAY=":1"```<br />
We have set up our display, but nothing is running.  It's because we havent started up anything with graphical output yet.  Once we start up QEMU, we can set up VNC Viewer.  To set up our VM, we use one command <br />

```> qemu-system-x86_64 boot_sect_1_4.bin```<br />
This starts the VM.  And we will now swtitch over to our VNC Viewer app.  We are going to create a new connection.  The first step is opening the app and pressing the green button.  The address will be 127.0.0.1:5901, and you can give it whatever name you want.  I prefer Termux because I use VNC Viewer for just LandOS.  127.0.0.1 is the localhost IP address and the port, 5901, is the port we will be connecting to.  If you are pushing your display to something other than process 1 (":1"), than the last number lf the port will be the process number.  For example:  You pushed your display to process 5(":5").  The last number in the port will be 5.  Thus, making the port 5905.  <br /><br />
Our VNC connection is now all set up.  Now all we need to do is connect to the display and enter your password.  Then, you can interact with the OS at your own will.<br /><br />
When you are finished you can simply close the VNC app, CTRL + C in Termux and type<br />

```> vncserver -kill :1 <----- Or whatever process it is ```<br /><br />

OPTIONAL:  Assembling the source code is rather simple.  First, navigate to the "src" directory.  Then, use the command:<br />

```> nasm boot_sect.asm -f bin -o boot_sect.bin```<br />
This creates the binary file you can run in QEMU.  Remember, this is completely optional.  As you are given an already assembled .bin file in the "images" directory.



changelog.txt
7/6/2022:
• LandOS changes background color, color can be changed
  by changing the value in bl. Tested values:
  - 0x01: Blue
  - 0x02: Green
  - 0x03: Sky Blue
  - 0x04: Red
  - 0x05: Pink
  - 0x06: Yellow
  - 0x07: Gray
  - 0x08: Dark Green
  - 0x09: Cobalt Blue
• New function that changes background color
• Fixed bug causing OS to print random characters
• WIP: Began working on a function to draw a new screen

7/10/2022:
• Finshed "newscreen" function
• Utilized "newscreen" function to clear ugly text after boot
• Set *X register to 0x00 if both *l and *h register's value was 0x00
• "keytocont" is now "to_main"
• Moved around some strings

7/11/2022:
• Added a safe return place ("main") for the OS to return to when finished with a program
• "to_main" brings you to "main"
• Changed a few calls to jumps
• OS is currently an infinite loop of jumping to "main" after key input

7/12/2022:
• LandOS's main screen now only responds to the follow scan codes:
  - 72 (Up key)
  - 80 (Down key)
  - 28 (Enter)
• After recieving one of the following keys you are sent to "to_main" where the os responds to any scan code

7/20/2022:
• Created "BaseFeatures" directory which will hold everything we need for the kernel
• Multiple files were created to sort the code so you dont want to rip your teeth out looking for a specific function

7/27/2022:
• Created "drivers.asm" and "pixel.asm" files
• Created function called 'switch_vmode' to check current video mode and switch the video mode
  -  If video mode is 0x03 (Text, 80x25, 16fore/8back color) switch video mode to 0x10
  -  If video mode is 0x10 (Graphics, 640x350, 16 color) switch video mode to 0x03
• Created function called 'write_pixel' to write a singular pixel to the screen
• Created function called 'loop_write_pixel' to make a diagnal blue line out of pixels
• After pressing any key the OS utilizes the 'loop_write_pixel'

7/31/2022:
• OS now has small delay when drawing each pixel

8/29/2022:
• Setting up new function that will allow us to draw custom rectangels.
• winArr holds 4 values: length(first), width(second), currentLoc(third), and process(fourth)
  - These values will constantly be modified throught the program and will look completely different when it finishes
  - You will still have to change color and starting position in assembly code

src/kernel_op.asm
[BITS 16]

main:
call newscreen
mov si, welcomemsg
call printstring
call newline
mov si, keyinput
call printstring
call newline
jmp .nav_main
.nav_main: ; Moving cursor and selecting things on the main screen
; Get key and check if scan code, stored in ah, is Up(72), Down(80), or Enter(28)
mov ah, 0x00
int 0x16
cmp ah, 72 ; up
je to_main
cmp ah, 80 ; down
je to_main
cmp ah, 28 ; enter
je to_main
jmp .nav_main

printchar:
mov ah, 0x0e
mov al, bl
int 10h
ret

printstring:
mov ah, 0x0e
.repeat:
lodsb
cmp al, 0
je .done
int 10h
jmp .repeat
.done:
xor si, si ; clear out si
ret

newline:
pusha
mov ah,0x0e
mov al, 0DH
mov bl, al
int 0x10
mov al, 0AH
mov bl, al
int 0x10
popa
ret

; jump to main instead of calling because we will not need to return anytime soon
to_main:
call newscreen
mov si, returnmsg
call printstring
call newline
mov si, keyinput
call printstring
call newline
mov ah, 0x00
int 0x16
cmp al, 0
jne .cont
.cont:
mov si, test_msg
call printstring
call newline
jmp main

newscreen:
pusha
; Set background-color to blue
mov ah, 0xB
mov bh, 0x00
mov bl, 0x01
int 0x10
; Move cursor to top left
mov ah, 0x02
mov bh, 0x00
mov dx, 0x00
int 0x10
; Wipe the Screen
mov ah, 0x06
mov al, 0x00
mov bh, 0x07
mov cx, 0x00
mov dh, 0x24
mov dl, 0x79
int 0x10
popa
ret

changebgcolor:
mov ah, 0xB
mov bh, 0x00
mov bl, 0x01 ; Modify value here to change background color
int 0x10
ret

welcomemsg db 'Welcome to LandOS', 0
returnmsg db 'Returning to home page', 0
successmsg db 'Kernel successfully loaded', 0
keyinput db 'Press any key to continue...', 0
test_msg db 'Success', 0